### ISSUE TITLE
Register Parameter names in Gpflow

### ISSUE THREAD

**Issue by jmaronas** [GPflow/GPflow#1702]
# Feature request

When creating a Parameter, e.g an Inducing Point, gpflow sets tf.Variable( value, name = None) name to None. In this way we cannot track parameters later on by looping over model.trainable_variables.

This is clearly a tensorflow problem, since in my opinion it should define a name automatically. For example in pytorch if name is not specified then Pytorch uses class name (nn.Module) to set a name for parameters.

## Motivation

Track parameters. An example could be the case in which you want to use two different optimizers for different parameters. For example in Deep Kernel Learning we might wanna use SGD for the Neural Network (since SGD usually performs better on these models) and then use Adam for the GP part. 

Also another motivation would be to use different learning rates or optimization schedulings ( for example natural gradients for variational parameters and LBGFS for hyperparameters)

Another motivation could be for setting parameters to be non trainable without having to reference them directly (i.e can be done in a loop filtering by name).

The only way we can do this at the moment is by directly adding the parameters to a list manually, i.e variable_list.append( model.kernel.trainable_variables ), which in my opinion is tedious as we filtering by name in a loop seems easier and more dynamic.

## Solution:

The solution is quite simple. For example in the svgp code we can just replace:

```python
q_mu = np.zeros((num_inducing, self.num_latent_gps)) if q_mu is None else q_mu
        self.q_mu = Parameter(q_mu, dtype=default_float())  # [M, P]

        if q_sqrt is None:
            if self.q_diag:
                ones = np.ones((num_inducing, self.num_latent_gps), dtype=default_float())
                self.q_sqrt = Parameter(ones, transform=positive())  # [M, P]
            else:
                q_sqrt = [
                    np.eye(num_inducing, dtype=default_float()) for _ in range(self.num_latent_gps)
                ]
                q_sqrt = np.array(q_sqrt)
                self.q_sqrt = Parameter(q_sqrt, transform=triangular())  # [P, M, M]
        else:
            if q_diag:
                assert q_sqrt.ndim == 2
                self.num_latent_gps = q_sqrt.shape[1]
                self.q_sqrt = Parameter(q_sqrt, transform=positive())  # [M, L|P]
            else:
                assert q_sqrt.ndim == 3
                self.num_latent_gps = q_sqrt.shape[0]
                num_inducing = q_sqrt.shape[1]
                self.q_sqrt = Parameter(q_sqrt, transform=triangular())  # [L|P, M, M]

```

to 

```python
q_mu = np.zeros((num_inducing, self.num_latent_gps)) if q_mu is None else q_mu
        self.q_mu = Parameter(q_mu, dtype=default_float(), name = 'svgp_variational_mean')  # [M, P]

        if q_sqrt is None:
            if self.q_diag:
                ones = np.ones((num_inducing, self.num_latent_gps), dtype=default_float())
                self.q_sqrt = Parameter(ones, transform=positive(), name = 'svgp_variational_covariance')  # [M, P]
            else:
                q_sqrt = [
                    np.eye(num_inducing, dtype=default_float()) for _ in range(self.num_latent_gps)
                ]
                q_sqrt = np.array(q_sqrt)
                self.q_sqrt = Parameter(q_sqrt, transform=triangular(),  name = 'svgp_variational_covariance')  # [P, M, M]
        else:
            if q_diag:
                assert q_sqrt.ndim == 2
                self.num_latent_gps = q_sqrt.shape[1]
                self.q_sqrt = Parameter(q_sqrt, transform=positive(),  name = 'svgp_variational_covariance')  # [M, L|P]
            else:
                assert q_sqrt.ndim == 3
                self.num_latent_gps = q_sqrt.shape[0]
                num_inducing = q_sqrt.shape[1]
                self.q_sqrt = Parameter(q_sqrt, transform=triangular(), name = 'svgp_variational_covariance')  # [L|P, M, M]

```

Note that at the moment Gpflow allows the user to pass in a name argument when creating different instances like kernels, or inducing points. But then these names are not use to register the parameters, which in my opinion is something we could fix for a better usage of Gpflow.

**Comment by st--** [comment]
This would be great, and in fact GPflow 1 had its own parameter-tracking system that did give names by default, but I think it'd be a significant amount of work to get this working in a consistent way across all of GPflow... would probably require a bit of upfront discussion how to actually do it.
